core/src/test/java/com/graphhopper/GraphHopperAPITest.java

```diff
@@ -115,4 +124,69 @@ public void testNoLoad() {
             assertTrue(ex.getMessage(), ex.getMessage().startsWith("Do a successful call to load or importOrLoad before routing"));
         }
     }
+
+    @Test
+    public void testConcurrentGraphChange() throws InterruptedException {
+        final GraphHopperStorage graph = new GraphBuilder(encodingManager).create();
+        initGraph(graph);
+        graph.edge(1, 2, 10, true);
+
+        final AtomicInteger checkPointCounter = new AtomicInteger(0);
+        final GraphHopper graphHopper = new GraphHopper() {
+            @Override
+            protected ChangeGraphHelper createChangeGraphHelper(Graph graph, LocationIndex locationIndex) {
+                return new ChangeGraphHelper(graph, locationIndex) {
+                    @Override
+                    public long applyChanges(EncodingManager em, Collection<JsonFeature> features) {
+                        // force sleep inside the lock
+                        try {
+                            Thread.sleep(1000);
+                        } catch (InterruptedException e) {
+                            throw new RuntimeException(e);
+                        }
+                        checkPointCounter.incrementAndGet();
+                        return super.applyChanges(em, features);
+                    }
+                };
+            }
+        }.setStoreOnFlush(false).setEncodingManager(encodingManager).setCHEnabled(false).
+                loadGraph(graph);
+
+        GHResponse rsp = graphHopper.route(new GHRequest(42, 10.4, 42, 10));
+        assertFalse(rsp.toString(), rsp.hasErrors());
+        assertEquals(1800, rsp.getBest().getTime());
+
+        final List<JsonFeature> list = new ArrayList<>();
+        Map<String, Object> properties = new HashMap<>();
+        properties.put("speed", 5);
+
+        list.add(new JsonFeature("1", "bbox",
+                new BBox(10.399, 10.4, 42.0, 42.001),
+                null, properties));
+
+        Thread thread1 = new Thread() {
+            @Override
+            public void run() {
```

Comment:
I would say we want a test which fails when we comment-out the locking, and succeeds otherwise.

I don't mind your 1-second-sleep. I have no better idea myself at the moment. The problem I described is unrelated to that.